### PR TITLE
Moveit generic

### PR DESCRIPTION
**Pull request by Nikhil-Singhal-06**
### Pull request checklist

- [ ] [Issue](https://github.com/IntelLabs/scenario_execution/issues) created that explains the change and why it's needed
- [ ] Tests are part of the PR (for bug fixes / features)
- [ ] [Docs](https://intellabs.github.io/scenario_execution/) reviewed and added / updated if needed (for bug fixes / features)
- [ ] Format and Lint checks (`make check`) pass locally
- [ ] PR applies Apache 2.0 License to all code files

### Pull request type

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation changes
- [ ] Other (please describe)

### CODE REVIEW

File: simulation/gazebo/arm_sim_scenario/launch/sim_moveit_scenario_launch.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
+# Copyright (C) 2024 Intel Corporation
```
Comment (fmirus) [inline simulation/gazebo/arm_sim_scenario/launch/sim_moveit_scenario_launch.py:1]: could you please add an rviz-file to this package for easier testing/visualization?

**Review by fmirus — CHANGES_REQUESTED** [review]

File: examples/example_moveit/example_moveit.osc [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+        keep(it.arm_joints == ['panda_joint1','panda_joint2','panda_joint3','panda_joint4','panda_joint5','panda_joint6','panda_joint7'])
+        keep(it.gripper_joints == ['panda_finger_joint1'])
+        keep(it.arm_group == 'panda_arm')
+        keep(it.gripper_group == 'hand')
+    do serial:
+        manipulator.move_to_joint_pose(
+            goal_pose: ['-2.82', '+1.01', '-2.40', '-1.46', '0.57', '2.47', '0.0'],
+            move_group: move_group_type!arm
+        )
+        wait elapsed(3s)
```
[the remark this answers is not in the document]
Comment (Nikhil-Singhal-06) [reply examples/example_moveit/example_moveit.osc]: Not necessary, I just added for visualization.

File: .gitmodules [hunk 3]
```diff
@@ -1,3 +1,7 @@
 [submodule "dependencies/py_trees_ros"]
 	path = dependencies/py_trees_ros
 	url = https://github.com/splintered-reality/py_trees_ros.git
+[submodule "dependencies/moveit_resources"]
```
Comment (fred-labs) [inline .gitmodules:4]: to be able to submit to the ros debian repositories, we try to get rid of submodules. Is this available as debian package?

File: examples/example_moveit/example_moveit.osc [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    timeout(60s)
+    manipulator: arm with:
+        keep(it.arm_joints == ['panda_joint1','panda_joint2','panda_joint3','panda_joint4','panda_joint5','panda_joint6','panda_joint7'])
+        keep(it.gripper_joints == ['panda_finger_joint1','panda_finger_joint2'])
+        keep(it.arm_group == 'panda_arm')
+        keep(it.gripper_group == 'hand')
+        keep(it.end_effector == 'panda_hand')
+        keep(it.base_link == 'panda_link0')
+    do serial:
+        joint_pose: serial:
```
Comment (fred-labs) [inline examples/example_moveit/example_moveit.osc:15]: you should be able to get rid of the "serial". Just add "joint_pose:" in front of the action call

File: libs/scenario_execution_moveit/package.xml [hunk 5]
```diff
@@ -0,0 +1,24 @@
+<?xml version="1.0"?>
+<?xml-model href="http://download.ros.org/schema/package_format3.xsd" schematypens="http://www.w3.org/2001/XMLSchema"?>
+<package format="3">
+  <name>scenario_execution_moveit</name>
```
Comment (fred-labs) [inline libs/scenario_execution_moveit/package.xml:4]: isn't it "moveit2"?

**Review by fred-labs — CHANGES_REQUESTED** [review]

File: libs/scenario_execution_moveit/package.xml (hunk 5 above)
Comment (Nikhil-Singhal-06) [reply libs/scenario_execution_moveit/package.xml:4]: Yes the lib is called moveit2 but the debian packages are still ros-distor-moveit and moveit_msgs.

Should I still rename the lib?

File: examples/example_moveit2/example_moveit2.osc [hunk 6]
```diff
@@ -0,0 +1,27 @@
+import osc.helpers
+import osc.ros
+import osc.moveit2
+
+scenario example_moveit2:
+    timeout(60s)
+    manipulator: arm with:
```
Comment (fred-labs) [inline examples/example_moveit2/example_moveit2.osc:7]: please do not use "keep". Instead use arm(arm_joints: ....., gripper_joints: ....)

File: libs/scenario_execution_moveit2/scenario_execution_moveit2/lib_osc/moveit2.osc [hunk 7]
```diff
@@ -0,0 +1,38 @@
+import osc.standard.base
+import osc.robotics
+
+enum move_group_type: [
+  arm,
+  gripper
+]
+
+actor arm inherits robot:
```
Comment (fred-labs) [inline libs/scenario_execution_moveit2/scenario_execution_moveit2/lib_osc/moveit2.osc:9]: please add descriptions. Follow other osc library files (at least one line describing the action itself. and after each argument, add a description for that. Argument indentation of other files is 42 characters)

File: simulation/gazebo/arm_sim_scenario/launch/moveit_launch.py [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+            mappings={
+                "ros2_control_hardware_type": ros2_control_hardware_type
+            },
+        )
+        .robot_description_semantic(file_path="config/panda.srdf")
+        .trajectory_execution(file_path="config/gripper_moveit_controllers.yaml")
+        .planning_pipelines(
+            pipelines=["ompl", "chomp", "pilz_industrial_motion_planner"]
+        )
+        # .planning_scene_monitor(publish_robot_description=True)
```
Comment (fred-labs) [inline simulation/gazebo/arm_sim_scenario/launch/moveit_launch.py:57]: cleanup

**Review by fred-labs — CHANGES_REQUESTED** [review]

File: libs/scenario_execution_moveit2/scenario_execution_moveit2/lib_osc/moveit2.osc [hunk 9]
```diff
@@ -0,0 +1,48 @@
+import osc.standard.base
```
Comment (fred-labs) [inline libs/scenario_execution_moveit2/scenario_execution_moveit2/lib_osc/moveit2.osc:1]: please add to docs/libraries.rst

**Review by fred-labs — CHANGES_REQUESTED** [review]

File: docs/libraries.rst [hunk 10]
```diff
@@ -1454,3 +1456,143 @@ Capture the screen content within a video.
      - ``float``
      - ``25.0``
      - Frame-rate of the resulting video
+
+Moveit2
+-------
+
+The library contains actions to interact with the `Moveit2 <https://moveit.picknik.ai/main/index.html>`__ manipulation stack. Import it with ``import osc.moveit2``. It is provided by the package :repo_link:`libs/scenario_execution_moveit2`.
+
+``arm``
```
Comment (fred-labs) [inline docs/libraries.rst:1465]: please add heading "Actors".

File: docs/libraries.rst [hunk 11]
```diff
@@ -1454,3 +1456,143 @@ Capture the screen content within a video.
...
+     - ``string``
+     -
+     - Name of the end effector component (e.g., hand or tool)
+   * - ``base_link``
+     - ``string``
+     -
+     - Name of the robot's base link for reference in kinematics
+
+``arm.move_to_joint_pose()``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
Comment (fred-labs) [inline docs/libraries.rst:1508]: reduce line-characters to match text

File: docs/libraries.rst [hunk 12]
```diff
@@ -1454,3 +1456,143 @@ Capture the screen content within a video.
...
+
+Use Moveit2 to move the arm joints to specified joint positions.
```
Comment (fred-labs) [inline docs/libraries.rst:1510]: please clarify which action is called. Is there some additional documentation available?

File: docs/libraries.rst [hunk 13]
```diff
@@ -1454,3 +1456,143 @@ Capture the screen content within a video.
...
+     -
+     - List joint positions to move to
+   * - ``move_group``
+     - ``move_group_type``
+     -
+     - Move group type. Allowed [arm, gripper] (e.g. ``[move_group_type!arm, move_group_type!gripper]``)
+   * - ``plan_only``
+     - ``bool``
+     - ``false``
+     - If true, the action returns an executable plan in the response but does not attempt execution
```
Comment (fred-labs) [inline docs/libraries.rst:1532]: is the action really returning the plan? where is it stored?

File: docs/libraries.rst [hunk 14]
```diff
@@ -1454,3 +1456,143 @@ Capture the screen content within a video.
...
+     - ``true``
+     - If true, replan if plan becomes invalidated during execution
+   * - ``tolerance``
+     - ``float``
+     - ``0.001``
+     - The acceptable range of variation around both the start and goal positions.
+   * - ``max_velocity_scaling_factor``
+     - ``float``
+     - ``false``
+     - Scaling factors for optionally reducing the maximum joint velocities
```
Comment (fred-labs) [inline docs/libraries.rst:1544]: false seems like an invalid value.

File: docs/libraries.rst [hunk 15]
```diff
@@ -1454,3 +1456,143 @@ Capture the screen content within a video.
...
+Use moveit2 to move the end-effector to specified position
+
+   * - Parameter
+     - Type
+     - Default
+     - Description
+   * - ``goal_pose``
+     - ``list of string``
+     -
+     - end effector pose to move to ``[x, y, z, quatx, quaty, quatz, w]``
```
Comment (fred-labs) [inline docs/libraries.rst:1570]: please use pose_3d

File: docs/libraries.rst [hunk 16]
```diff
@@ -1454,3 +1456,143 @@ Capture the screen content within a video.
...
+   * - ``plan_only``
+     - ``bool``
+     - ``false``
+     - If true, the action returns an executable plan in the response but does not attempt execution
```
Comment (fred-labs) [inline docs/libraries.rst:1574]: same as above

File: docs/libraries.rst [hunk 17]
```diff
@@ -1454,3 +1456,143 @@ Capture the screen content within a video.
...
+     - ``bool``
+     - ``true``
+     - If true, replan if plan becomes invalidated during execution
+   * - ``tolerance``
+     - ``float``
+     - ``0.001``
+     - The acceptable range of variation around both the start and goal positions.
+   * - ``max_velocity_scaling_factor``
+     - ``float``
+     - ``false``
```
Comment (fred-labs) [inline docs/libraries.rst:1585]: same as above

File: docs/libraries.rst [hunk 18]
```diff
@@ -1454,3 +1456,143 @@ Capture the screen content within a video.
...
+     - Scaling factors for optionally reducing the maximum joint velocities
+   * - ``namespace_override``
+     - ``string``
+     - ``false``
+     - if set, it's used as namespace (instead of the associated actor's name)
+   * - ``action_topic``
+     - ``string``
+     - ``move_action``
```
Comment (fred-labs) [inline docs/libraries.rst:1593]: is this really the same name as for the other action?!

**Review by fred-labs — CHANGES_REQUESTED** [review]

File: docs/libraries.rst (hunk 18 above)
Comment (Nikhil-Singhal-06) [reply docs/libraries.rst:1593]: yes

File: docs/libraries.rst (hunk 13 above)
Comment (Nikhil-Singhal-06) [reply docs/libraries.rst:1532]: yes, it returns the plan as action response. as stated here 

https://docs.ros.org/en/melodic/api/moveit_msgs/html/msg/PlanningOptions.html

Comment (fred-labs) [reply docs/libraries.rst:1532]: from a scenario point of view, there is no returned value. right?

Comment (Nikhil-Singhal-06) [reply docs/libraries.rst:1532]: right, if the target position is valid then scenario will return success. Also one can visualize the planning in Rviz, the arm shadow will follow the path but arm will not get executed (actual position will not change).

Comment (fred-labs) [reply docs/libraries.rst:1535]: then something like this should describe it better (feel free to improve)
If true, the plan is not executed after calculation. Calculated plan can be visualized in rviz.

Comment (Nikhil-Singhal-06) [reply docs/libraries.rst:1535]: Sure

File: docs/libraries.rst [hunk 19]
```diff
@@ -1454,3 +1456,146 @@ Capture the screen content within a video.
...
+Moveit2
+-------
+
+The library contains actions to interact with the `Moveit2 <https://moveit.picknik.ai/main/index.html>`__ manipulation stack. Import it with ``import osc.moveit2``. It is provided by the package :repo_link:`libs/scenario_execution_moveit2`.
+
+Actors
+^^^^^^
+
+``arm``
+^^^^^^^
```
Comment (fred-labs) [inline docs/libraries.rst:1469]: wrong section type. use """"""""" instead

File: docs/libraries.rst [hunk 20]
```diff
@@ -1454,3 +1456,146 @@ Capture the screen content within a video.
...
+     - Name of the end effector component (e.g., hand or tool)
+   * - ``base_link``
+     - ``string``
+     -
+     - Name of the robot's base link for reference in kinematics
+
+``arm.move_to_joint_pose()``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Use MoveIt2 to move the arm joints to specified joint positions, utilizing `MoveGroup action <https://docs.ros.org/en/noetic/api/moveit_msgs/html/action/MoveGroup.html>`__ from the move_group node () by specifying target joint values.
```
Comment (fred-labs) [inline docs/libraries.rst:1513]: is there a non-rosdistro-specific link? noetic seems rather old

Comment (fred-labs) [reply docs/libraries.rst:1513]: what does "()" mean?

File: docs/libraries.rst (hunk 15 above)
Comment (fred-labs) [reply docs/libraries.rst:1573]: invalid description

File: examples/example_moveit2/example_moveit2.osc [hunk 21]
```diff
@@ -0,0 +1,26 @@
...
+    timeout(60s)
+    manipulator: arm = arm(arm_joints: ['panda_joint1','panda_joint2','panda_joint3','panda_joint4','panda_joint5','panda_joint6','panda_joint7'],
+        gripper_joints: ['panda_finger_joint1','panda_finger_joint2'],
+        arm_group: 'panda_arm',
+        gripper_group: 'hand',
+        end_effector: 'panda_hand',
+        base_link: 'panda_link0')
+    do serial:
+        joint_pose: manipulator.move_to_joint_pose(
+                goal_pose: ['-2.82', '+1.01', '-2.40', '-1.46', '0.57', '2.47', '0.0'],
```
Comment (fred-labs) [inline examples/example_moveit2/example_moveit2.osc:15]: use pose_3d.

pose_3d(position_3d(-2.82, 1.01, -2.4), orientation_3d(roll, pitch, yaw))

File: examples/example_moveit2/example_moveit2.osc [hunk 22]
```diff
@@ -0,0 +1,26 @@
...
+                move_group: move_group_type!arm)
+        open_gripper: manipulator.move_to_joint_pose(
+                goal_pose: ['0.04'],
```
Comment (fred-labs) [inline examples/example_moveit2/example_moveit2.osc:18]: use numbers

File: libs/scenario_execution_moveit2/scenario_execution_moveit2/lib_osc/moveit2.osc [hunk 23]
```diff
@@ -0,0 +1,48 @@
...
+    base_link: string                     # Name of the robot's base link for reference in kinematics
+
+
+###########
+# Actions
+###########
+
+action arm.move_to_joint_pose:
+    # Use Moveit2 to move the arm joints to specified joint positions
+    goal_pose: list of string            # list joint positions to move to
```
Comment (fred-labs) [inline libs/scenario_execution_moveit2/scenario_execution_moveit2/lib_osc/moveit2.osc:29]: update

File: simulation/gazebo/arm_sim_scenario/package.xml [hunk 24]
```diff
@@ -0,0 +1,33 @@
+<?xml version="1.0"?>
+<?xml-model href="http://download.ros.org/schema/package_format3.xsd" schematypens="http://www.w3.org/2001/XMLSchema"?>
+<package format="3">
+  <name>arm_sim_scenario</name>
+  <version>1.2.0</version>
+  <description>Moveit2 Arm Simulation Scenario Execution</description>
```
Comment (fred-labs) [inline simulation/gazebo/arm_sim_scenario/package.xml:6]: MoveIt2

**Review by fred-labs — CHANGES_REQUESTED** [review]

File: examples/example_moveit2/example_moveit2.osc (hunk 21 above)
Comment (Nikhil-Singhal-06) [reply examples/example_moveit2/example_moveit2.osc:15]: For joint to pose we cannot use pose3d as the values are the position of each joint and the list size can vary depending on number of joints.

Comment (fred-labs) [reply examples/example_moveit2/example_moveit2.osc:15]: got it.....but at least list of floats should be possible, right?

**Review by fred-labs — APPROVED** [review]

**Review by fmirus — APPROVED** [review]

**Pull request merged**